### PR TITLE
feat(conditions): allow checks condition in and/or

### PR DESCRIPTION
**Pull request by sileht**
Change-Id: I9f2df8ae19a41352a39446e2e114a60c1ce64292

### CODE REVIEW

**Comment by sileht** [comment]
This pull request is part of a stack:
* feat(rules): allow and/or in conditions ([#2691](https://github.com/Mergifyio/mergify-engine/pull/2691))
* feat: allow merged/closed/author/base/head in and/or conditions ([#2692](https://github.com/Mergifyio/mergify-engine/pull/2692))
* feat(conditions): allow checks condition in and/or ([#2693](https://github.com/Mergifyio/mergify-engine/pull/2693))

File: mergify_engine/actions/merge_base.py [hunk 1]
```diff
@@ -99,6 +99,64 @@ def strict_merge_parameter(v):
...
+    # NOTE(sileht): All checks have reported their status ?
+    await conditions_with_all_checks(ctxt.pull_request)
+    ctxt.log.debug(
+        "are checks reported their status? %s",
+        conditions_with_all_checks.get_summary(),
+    )
+    if not conditions_with_all_checks.match:
+        return check_api.Conclusion.PENDING
+
+    # NOTE(sileht): Are remaing unmatch checks pending ?
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:148]:
```suggestion
    # NOTE(sileht): Are remaining unmatch checks pending ?
```

File: mergify_engine/context.py [hunk 2]
```diff
@@ -815,7 +815,7 @@ async def _get_consolidated_data(self, name: str) -> ContextAttributeType:
                 for ctext, state in (await self.checks).items()
                 if state == "skipped"
             ]
-        elif name == "check-pending":
+        elif name in ["check", "check-pending"]:
```
Comment (jd) [inline mergify_engine/context.py:818]:
```suggestion
        elif name in ("check", "check-pending"):
```

File: mergify_engine/rules/__init__.py [hunk 3]
```diff
@@ -72,6 +72,22 @@ def update(self, condition: typing.Union[str, FakeTreeT]) -> None:
                 error_message=str(e),
             )
 
+    def update_attribute_name(self, new_name: str) -> None:
+        negate = False
+        tree = typing.cast(filter.TreeT, self.partial_filter.tree)
+        if "-" in tree:
+            negate = True
```
Comment (jd) [inline mergify_engine/rules/__init__.py:79]:
```suggestion
        tree = typing.cast(filter.TreeT, self.partial_filter.tree)
        negate = "-" in tree
```

File: mergify_engine/actions/merge_base.py [hunk 4]
```diff
@@ -99,6 +99,64 @@ def strict_merge_parameter(v):
...
+        rule.conditions.walk(),
+        conditions_without_checks.walk(),
+        conditions_with_all_checks.walk(),
+        conditions_with_check_pending.walk(),
+    ):
+        attr = evaluated_condition.get_attribute_name()
+        if attr.startswith("check-") or attr.startswith("status-"):
+            condition_without_check.update("number>0")
+            condition_with_check_pending.update_attribute_name("check-pending")
+            condition_with_all_check.update_attribute_name("check")
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:128]: Do we need to handle `status-` here too?

File: mergify_engine/actions/merge_base.py [hunk 5]
```diff
@@ -99,6 +99,64 @@ def strict_merge_parameter(v):
...
+
+    # NOTE(sileht): Something unrelated to checks unmatch ?
+    await conditions_without_checks(ctxt.pull_request)
+    ctxt.log.debug(
+        "does something unrealted to checks don't match? %s",
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:133]:
```suggestion
        "something unrelated to checks doesn't match? %s",
```

File: mergify_engine/actions/merge_base.py [hunk 6]
```diff
@@ -99,6 +99,64 @@ def strict_merge_parameter(v):
...
+        conditions_without_checks.get_summary(),
+    )
+    if not conditions_without_checks.match:
+        return check_api.Conclusion.FAILURE
+
+    # NOTE(sileht): All checks have reported their status ?
+    await conditions_with_all_checks(ctxt.pull_request)
+    ctxt.log.debug(
+        "are checks reported their status? %s",
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:142]:
```suggestion
        "did checks report their status? %s",
```

File: mergify_engine/actions/merge_base.py [hunk 7]
```diff
@@ -99,6 +99,64 @@ def strict_merge_parameter(v):
...
+    # NOTE(sileht): Something unrelated to checks unmatch ?
+    await conditions_without_checks(ctxt.pull_request)
+    ctxt.log.debug(
+        "does something unrealted to checks don't match? %s",
+        conditions_without_checks.get_summary(),
+    )
+    if not conditions_without_checks.match:
+        return check_api.Conclusion.FAILURE
+
+    # NOTE(sileht): All checks have reported their status ?
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:139]:
```suggestion
    # NOTE(sileht): Have all checks reported their status?
```

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/actions/merge_base.py (hunk 4 above)
Comment (sileht) [reply mergify_engine/actions/merge_base.py:128]: No, this replaces all check-XXX and status-XXX in `check`.

**Review by jd — APPROVED** [review]
@sileht still one typo in log if you want to fix it

**Pull request merged**